docs/source/configuring-pbft.rst
****************
Configuring PBFT
****************

Sawtooth PBFT is configured in two ways:

- Local settings - Command-line (CLI) options when running the PBFT executable
  or service

- Network settings - On-chain settings that are defined for the Sawtooth network


.. _cli-options-label:

=========================
PBFT Command-Line Options
=========================

Sawtooth PBFT uses command-line options to provide local configuration. These
options only affect the local behavior of a PBFT node, not the network as a
whole. The PBFT consensus engine, ``pbft-engine``, has the following
command-line options:

- | ``-C, --connect CONNECT``
  | (Optional; default ``tcp://localhost:5050``)
  | Connection endpoint for validator

- | ``-b, --exponential-retry-base BASE``
  | (Optional; default 100 ms)
  | Base timeout for exponential backoff used for validator requests

- | ``-m, --exponential-retry-max MAX``
  | (Optional; default 60000 ms)
  | Max timeout for exponential backoff used for validator requests

- | ``-L, --log-config LOG_CONFIG``
  | (Optional)
  | Path to logging config file; if not present, console logging is used

- | ``-l, --max-log-size MAX_LOG_SIZE``
  | (Optional; default 10000 messages)
  | How large the PBFT log is allowed to get before being pruned

- | ``-s, --storage-location STORAGE_LOCATION``
  | (Optional; default ``memory``)
  | Where to store PBFT's state: ``memory`` or ``disk+/path/to/file``

- | ``-u, --update-recv-timeout TIMEOUT``
  | (Optional; default 10 ms)
  | Timeout for receiving an update from the validator

- | ``-v``
  | (Optional)
  | Increase output verbosity


.. _on-chain-settings-label:

======================
PBFT On-Chain Settings
======================

Sawtooth PBFT includes on-chain settings for network-wide configuration on a
Hyperledger Sawtooth network. These settings affect how the whole network
operates, so it is desirable that they be the same on all nodes. The `Settings
transaction processor <https://sawtooth.hyperledger.org/docs/core/releases/latest/transaction_family_specifications/settings_transaction_family.html>`__
(or an equivalent) is required to process these settings.

.. tip::

   To display the existing settings, use `sawtooth settings
   list <https://sawtooth.hyperledger.org/docs/core/releases/latest/cli/sawtooth.html#sawtooth-settings-list>`__.

   To change a setting, use `sawset proposal
   create <https://sawtooth.hyperledger.org/docs/core/releases/latest/cli/sawset.html#sawset-proposal-create>`__.
   This command requires a signing key (the ``--key`` option) that specifies the
   public key of a user or validator that has permission to change settings. See
   ``sawtooth.identity.allowed_keys`` in `Configuring Validator and Transactor
   Permissions <https://sawtooth.hyperledger.org/docs/core/releases/latest/sysadmin_guide/configuring_permissions.html>`__.

- | ``sawtooth.consensus.pbft.block_publishing_delay``
  | (Optional; default 1000 ms)
  | How often to try to publish a block.

- | ``sawtooth.consensus.pbft.commit_timeout``
  | (Optional; default 10000 ms)
  | How long to wait (after Pre-Preparing) for the node to commit the block
  | before determining that the primary node is faulty.

- | ``sawtooth.consensus.pbft.forced_view_change_interval``
  | (Optional; default 100 blocks)
  | Number of blocks to commit before forcing a view change.

- | ``sawtooth.consensus.pbft.idle_timeout``
  | (Optional; default 30000 ms)
  | How long to wait for the next ``BlockNew`` and ``PrePrepare`` messages
  | before determining that the primary node is faulty. The idle timeout must be
  | longer than the block publishing delay.

- | ``sawtooth.consensus.pbft.members``
  | (Required)
  | List of validator public keys for the member nodes in the PBFT network,
  | as a comma-separated list (in a JSON-formatted string):
  | ``[public-key-1, public-key-2, ..., public-key-n]``

- | ``sawtooth.consensus.pbft.view_change_duration``
  | (Optional; default 5000 ms)
  | How long to wait for a valid ``NewView`` message before starting the next
  | view change. For more information, see :ref:`view-changing-mode-label`.


.. Licensed under Creative Commons Attribution 4.0 International License
.. https://creativecommons.org/licenses/by/4.0/


docs/source/on-chain-settings.rst
**********************
PBFT On-Chain Settings
**********************

Sawtooth PBFT includes on-chain settings for configuring PBFT consensus on a
Hyperledger Sawtooth network. The `Settings transaction
processor <https://sawtooth.hyperledger.org/docs/core/releases/latest/transaction_family_specifications/settings_transaction_family.html>`__
(or an equivalent) is required to process these settings.

.. tip::

   To display the existing settings, use `sawtooth settings
   list <https://sawtooth.hyperledger.org/docs/core/releases/latest/cli/sawtooth.html#sawtooth-settings-list>`__.

   To change a setting, use `sawset proposal
   create <https://sawtooth.hyperledger.org/docs/core/releases/latest/cli/sawset.html#sawset-proposal-create>`__.
   This command requires a signing key (the ``--key`` option) that specifies the
   public key of a user or validator that has permission to change settings. See
   ``sawtooth.identity.allowed_keys`` in `Configuring Validator and Transactor
   Permissions <https://sawtooth.hyperledger.org/docs/core/releases/latest/sysadmin_guide/configuring_permissions.html>`__.

- | ``sawtooth.consensus.pbft.block_publishing_delay``
  | (Optional; default 1000 ms)
  | How often to try to publish a block.

- | ``sawtooth.consensus.pbft.commit_timeout``
  | (Optional; default 10000 ms)
  | How long to wait between block commits before determining that the primary
  | node is faulty.

- | ``sawtooth.consensus.pbft.forced_view_change_period``
  | (Optional; default 100 blocks)
  | Number of blocks to commit before forcing a view change.

- | ``sawtooth.consensus.pbft.idle_timeout``
  | (Optional; default 30000 ms)
  | How long to wait for the next ``BlockNew`` and ``PrePrepare`` messages
  | before determining that the primary node is faulty. The idle timeout must be
  | longer than the block duration.

- | ``sawtooth.consensus.pbft.members``
  | (Required)
  | List of validator public keys for the member nodes in the PBFT network,
  | as a comma-separated list (in a JSON-formatted string):
  | ``[public-key-1, public-key-2, ..., public-key-n]``

- | ``sawtooth.consensus.pbft.view_change_duration``
  | (Optional; default 5000 ms)
  | How long to wait for a valid ``NewView`` message before starting the next
  | view change. For more information, see :ref:`view-changing-mode-label`.


.. Licensed under Creative Commons Attribution 4.0 International License
.. https://creativecommons.org/licenses/by/4.0/


src/config.rs
/*
 * Copyright 2018 Bitwise IO, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * -----------------------------------------------------------------------------
 */

//! Initial configuration for a PBFT node

use std::collections::HashMap;
use std::time::Duration;

use hex;
use sawtooth_sdk::consensus::{
    engine::{BlockId, PeerId},
    service::Service,
};
use serde_json;

use crate::timing::retry_until_ok;

/// Contains the initial configuration loaded from on-chain settings and local configuration. The
/// `members` list is required; all other settings are optional (defaults used in their absence)
#[derive(Debug)]
pub struct PbftConfig {
    // Members of the PBFT network
    pub members: Vec<PeerId>,

    /// How long to wait in between trying to publish blocks
    pub block_publishing_delay: Duration,

    /// How long to wait for an update to arrive from the validator
    pub update_recv_timeout: Duration,

    /// The base time to use for retrying with exponential backoff
    pub exponential_retry_base: Duration,

    /// The maximum time for retrying with exponential backoff
    pub exponential_retry_max: Duration,

    /// How long to wait for the next BlockNew + PrePrepare before determining primary is faulty
    /// Must be longer than block_publishing_delay
    pub idle_timeout: Duration,

    /// How long to wait (after Pre-Preparing) for the node to commit the block before starting a
    /// view change (guarantees liveness by allowing the network to get "unstuck" if it is unable
    /// to commit a block)
    pub commit_timeout: Duration,

    /// When view changing, how long to wait for a valid NewView message before starting a
    /// different view change
    pub view_change_duration: Duration,

    /// How many blocks to commit before forcing a view change for fairness
    pub forced_view_change_interval: u64,

    /// How large the PbftLog is allowed to get before being pruned
    pub max_log_size: u64,

    /// Where to store PbftState ("memory" or "disk+/path/to/file")
    pub storage_location: String,
}

impl PbftConfig {
    pub fn default() -> Self {
        PbftConfig {
            members: Vec::new(),
            block_publishing_delay: Duration::from_millis(1000),
            update_recv_timeout: Duration::from_millis(10),
            exponential_retry_base: Duration::from_millis(100),
            exponential_retry_max: Duration::from_millis(60000),
            idle_timeout: Duration::from_millis(30000),
            commit_timeout: Duration::from_millis(10000),
            view_change_duration: Duration::from_millis(5000),
            forced_view_change_interval: 100,
            max_log_size: 10000,
            storage_location: "memory".into(),
        }
    }

    /// Load configuration from on-chain Sawtooth settings.
    ///
    /// Configuration loads the following settings:
    /// + `sawtooth.consensus.pbft.members` (required)
    /// + `sawtooth.consensus.pbft.block_publishing_delay` (optional, default 1000 ms)
    /// + `sawtooth.consensus.pbft.idle_timeout` (optional, default 30000 ms)
    /// + `sawtooth.consensus.pbft.commit_timeout` (optional, default 10000 ms)
    /// + `sawtooth.consensus.pbft.view_change_duration` (optional, default 5000 ms)
    /// + `sawtooth.consensus.pbft.forced_view_change_interval` (optional, default 100 blocks)
    ///
    /// # Panics
    /// + If block publishing delay is greater than the idle timeout
    /// + If the `sawtooth.consensus.pbft.members` setting is not provided or is invalid
    pub fn load_settings(&mut self, block_id: BlockId, service: &mut Service) {
        debug!("Getting on-chain settings for config");
        let settings: HashMap<String, String> = retry_until_ok(
            self.exponential_retry_base,
            self.exponential_retry_max,
            || {
                service.get_settings(
                    block_id.clone(),
                    vec![
                        String::from("sawtooth.consensus.pbft.members"),
                        String::from("sawtooth.consensus.pbft.block_publishing_delay"),
                        String::from("sawtooth.consensus.pbft.idle_timeout"),
                        String::from("sawtooth.consensus.pbft.commit_timeout"),
                        String::from("sawtooth.consensus.pbft.view_change_duration"),
                        String::from("sawtooth.consensus.pbft.forced_view_change_interval"),
                    ],
                )
            },
        );

        // Get the on-chain list of PBFT members or panic if it is not provided; the network cannot
        // function without this setting, since there is no way of knowing which nodes are members.
        self.members = get_members_from_settings(&settings);

        // Get durations
        merge_millis_setting_if_set(
            &settings,
            &mut self.block_publishing_delay,
            "sawtooth.consensus.pbft.block_publishing_delay",
        );
        merge_millis_setting_if_set(
            &settings,
            &mut self.idle_timeout,
            "sawtooth.consensus.pbft.idle_timeout",
        );
        merge_millis_setting_if_set(
            &settings,
            &mut self.commit_timeout,
            "sawtooth.consensus.pbft.commit_timeout",
        );
        merge_millis_setting_if_set(
            &settings,
            &mut self.view_change_duration,
            "sawtooth.consensus.pbft.view_change_duration",
        );

        // Check to make sure block_publishing_delay < idle_timeout
        if self.block_publishing_delay >= self.idle_timeout {
            panic!(
                "Block publishing delay ({:?}) must be less than the idle timeout ({:?})",
                self.block_publishing_delay, self.idle_timeout
            );
        }

        // Get integer constants
        merge_setting_if_set(
            &settings,
            &mut self.forced_view_change_interval,
            "sawtooth.consensus.pbft.forced_view_change_interval",
        );
    }
}

fn merge_setting_if_set<T: ::std::str::FromStr>(
    settings_map: &HashMap<String, String>,
    setting_field: &mut T,
    setting_key: &str,
) {
    merge_setting_if_set_and_map(settings_map, setting_field, setting_key, |setting| setting)
}

fn merge_setting_if_set_and_map<U, F, T>(
    settings_map: &HashMap<String, String>,
    setting_field: &mut U,
    setting_key: &str,
    map: F,
) where
    F: Fn(T) -> U,
    T: ::std::str::FromStr,
{
    if let Some(setting) = settings_map.get(setting_key) {
        if let Ok(setting_value) = setting.parse() {
            *setting_field = map(setting_value);
        }
    }
}

fn merge_millis_setting_if_set(
    settings_map: &HashMap<String, String>,
    setting_field: &mut Duration,
    setting_key: &str,
) {
    merge_setting_if_set_and_map(
        settings_map,
        setting_field,
        setting_key,
        Duration::from_millis,
    )
}

/// Get the list of PBFT members as a Vec<PeerId> from settings
///
/// # Panics
/// + If the `sawtooth.consenus.pbft.members` setting is unset or invalid
pub fn get_members_from_settings<S: std::hash::BuildHasher>(
    settings: &HashMap<String, String, S>,
) -> Vec<PeerId> {
    let members_setting_value = settings
        .get("sawtooth.consensus.pbft.members")
        .expect("'sawtooth.consensus.pbft.members' is empty; this setting must exist to use PBFT");

    let members: Vec<String> = serde_json::from_str(members_setting_value).unwrap_or_else(|err| {
        panic!(
            "Unable to parse value at 'sawtooth.consensus.pbft.members' due to error: {:?}",
            err
        )
    });

    members
        .into_iter()
        .map(|s| {
            hex::decode(s).unwrap_or_else(|err| {
                panic!("Unable to parse PeerId from string due to error: {:?}", err)
            })
        })
        .collect()
}
